File: app/modules/billing/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class BillingMember(DashboardModel):
    id: str
    name: str
    email: str
    role: Literal["Owner", "Member"]
    seat_type: Literal["ChatGPT", "Codex"]
    date_added: str


class BillingCycle(DashboardModel):
    start: datetime
    end: datetime


class BillingAccount(DashboardModel):
    id: str
    domain: str
    plan_code: str
    plan_name: str
    subscription_status: Literal["trialing", "active", "past_due", "canceled", "expired"]
    entitled: bool
    payment_status: Literal["paid", "requires_action", "past_due", "unpaid"]
    billing_cycle: BillingCycle
    renewal_at: datetime | None = None
    chatgpt_seats_in_use: int = Field(ge=0)
    codex_seats_in_use: int = Field(ge=0)
    members: list[BillingMember]


class BillingAccountsResponse(DashboardModel):
    accounts: list[BillingAccount]


class BillingAccountsUpdateRequest(DashboardModel):
    accounts: list[BillingAccount]


class BillingAccountCreateRequest(DashboardModel):
    domain: str = Field(min_length=1, max_length=255)
    plan_code: str = Field(default="business", min_length=1, max_length=64)
    plan_name: str = Field(default="Business", min_length=1, max_length=128)
    subscription_status: Literal["trialing", "active", "past_due", "canceled", "expired"] = "active"
    payment_status: Literal["paid", "requires_action", "past_due", "unpaid"] = "paid"
    entitled: bool = True
    renewal_at: datetime | None = None
    chatgpt_seats_in_use: int = Field(default=0, ge=0)
    codex_seats_in_use: int = Field(default=0, ge=0)

File: apps/frontend/src/features/billing/schemas.ts
import { z } from "zod";

export const BillingMemberSchema = z.object({
  id: z.string().min(1),
  name: z.string(),
  email: z.string(),
  role: z.enum(["Owner", "Member"]),
  seatType: z.enum(["ChatGPT", "Codex"]),
  dateAdded: z.string(),
});

export const BillingCycleSchema = z.object({
  start: z.coerce.date(),
  end: z.coerce.date(),
});

export const BillingAccountSchema = z.object({
  id: z.string().min(1),
  domain: z.string().min(1),
  planCode: z.string().min(1),
  planName: z.string().min(1),
  subscriptionStatus: z.enum(["trialing", "active", "past_due", "canceled", "expired"]),
  entitled: z.boolean(),
  paymentStatus: z.enum(["paid", "requires_action", "past_due", "unpaid"]),
  billingCycle: BillingCycleSchema,
  renewalAt: z.coerce.date().nullable(),
  chatgptSeatsInUse: z.number().int().nonnegative(),
  codexSeatsInUse: z.number().int().nonnegative(),
  members: z.array(BillingMemberSchema),
});

export const BillingAccountsResponseSchema = z.object({
  accounts: z.array(BillingAccountSchema),
});

export const BillingAccountsUpdateRequestSchema = z.object({
  accounts: z.array(BillingAccountSchema),
});

export const BillingAccountCreateRequestSchema = z.object({
  domain: z.string().min(1),
  planCode: z.string().min(1).default("business"),
  planName: z.string().min(1).default("Business"),
  subscriptionStatus: z
    .enum(["trialing", "active", "past_due", "canceled", "expired"])
    .default("active"),
  paymentStatus: z.enum(["paid", "requires_action", "past_due", "unpaid"]).default("paid"),
  entitled: z.boolean().default(true),
  renewalAt: z.coerce.date().nullable().optional(),
  chatgptSeatsInUse: z.number().int().nonnegative().default(0),
  codexSeatsInUse: z.number().int().nonnegative().default(0),
});

export type BillingMember = z.infer<typeof BillingMemberSchema>;
export type BillingCycle = z.infer<typeof BillingCycleSchema>;
export type BillingAccount = z.infer<typeof BillingAccountSchema>;
export type BillingAccountsResponse = z.infer<typeof BillingAccountsResponseSchema>;
export type BillingAccountsUpdateRequest = z.infer<typeof BillingAccountsUpdateRequestSchema>;
export type BillingAccountCreateRequest = z.infer<typeof BillingAccountCreateRequestSchema>;
